modules/apps/document-library/document-library-service/src/main/java/com/liferay/document/library/internal/configuration/admin/service/DLPreviewLimitManagedServiceFactory.java

```diff
@@ -0,0 +1,514 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2000 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.document.library.internal.configuration.admin.service;
+
+import com.liferay.document.library.exception.DLPreviewLimitConfigurationException;
+import com.liferay.document.library.internal.configuration.DLPreviewLimitConfiguration;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.configuration.metatype.annotations.ExtendedObjectClassDefinition;
+import com.liferay.portal.configuration.metatype.bnd.util.ConfigurableUtil;
+import com.liferay.portal.kernel.model.CompanyConstants;
+import com.liferay.portal.kernel.model.Group;
+import com.liferay.portal.kernel.model.GroupConstants;
+import com.liferay.portal.kernel.security.auth.CompanyThreadLocal;
+import com.liferay.portal.kernel.service.GroupLocalService;
+import com.liferay.portal.kernel.util.GetterUtil;
+import com.liferay.portal.kernel.util.HashMapDictionary;
+import com.liferay.portal.kernel.util.HashMapDictionaryBuilder;
+
+import java.util.Dictionary;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Supplier;
+
+import org.osgi.framework.Constants;
+import org.osgi.service.cm.Configuration;
+import org.osgi.service.cm.ConfigurationAdmin;
+import org.osgi.service.cm.ConfigurationException;
+import org.osgi.service.cm.ManagedServiceFactory;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Modified;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Marco Galluzzi
+ */
+@Component(
+	configurationPid = "com.liferay.document.library.internal.configuration.DLPreviewLimitConfiguration",
+	property = Constants.SERVICE_PID + "=com.liferay.document.library.internal.configuration.DLPreviewLimitConfiguration.scoped",
+	service = {
+		DLPreviewLimitManagedServiceFactory.class, ManagedServiceFactory.class
+	}
+)
+public class DLPreviewLimitManagedServiceFactory
```

Comment:
There's a ton of unnecessary indirection between this class and the provider. Make the provider impl implement both the provider interface and the ManagedServiceFactory. That way, the provider will have direct access to the configuration values and things will be much more simple.